be/src/util/arrow/starrocks_column_to_arrow.cpp

```diff
@@ -0,0 +1,308 @@
+struct ColumnToArrowConverter<PT, AT, is_nullable, ConvFloatAndIntegerGuard<PT, AT>> {
+    using StarRocksCppType = RunTimeCppType<PT>;
+    using StarRocksColumnType = RunTimeColumnType<PT>;
+    using ArrowType = ArrowTypeIdToType<AT>;
+    using ArrowCppType = ArrowTypeIdToCppType<AT>;
+    using ArrowBuilderType = typename arrow::TypeTraits<ArrowType>::BuilderType;
+    static inline arrow::Status convert(const ColumnPtr& column, arrow::MemoryPool* pool,
+                                        std::shared_ptr<arrow::Array>& array) {
+        DCHECK(!column->is_constant() && !column->only_null());
+        ArrowBuilderType builder(pool);
+        if constexpr (is_nullable) {
+            const auto* nullable_column = down_cast<NullableColumn*>(column.get());
+            const auto* data_column = down_cast<StarRocksColumnType*>(nullable_column->data_column().get());
+            const auto* null_column = down_cast<NullColumn*>(nullable_column->null_column().get());
+            const auto& data = data_column->get_data();
+            const auto num_rows = null_column->size();
+            for (auto i = 0; i < num_rows; ++i) {
+                if (nullable_column->is_null(i)) {
+                    ARROW_RETURN_NOT_OK(builder.AppendNull());
+                } else {
+                    ARROW_RETURN_NOT_OK(builder.Append(data[i]));
+                }
+            }
+        } else {
+            const auto* data_column = down_cast<StarRocksColumnType*>(column.get());
+            const auto& data = data_column->get_data();
+            ARROW_RETURN_NOT_OK(builder.AppendValues(data));
+        }
+        return builder.Finish(&array);
+    }
+};
+
+DEF_PRED_GUARD(ConvDecimalGuard, is_conv_decimal, PrimitiveType, PT, ArrowTypeId, AT)
+#define IS_CONV_DECIMAL_CTOR(PT, AT) DEF_PRED_CASE_CTOR(is_conv_decimal, PT, AT)
+#define IS_CONV_DECIMAL_R(AT, ...) DEF_BINARY_RELATION_ENTRY_SEP_SEMICOLON_R(IS_CONV_DECIMAL_CTOR, AT, ##__VA_ARGS__)
+
+IS_CONV_DECIMAL_R(ArrowTypeId::DECIMAL, TYPE_DECIMALV2, TYPE_DECIMAL32, TYPE_DECIMAL64, TYPE_DECIMAL128)
+
+template <PrimitiveType PT, ArrowTypeId AT, bool is_nullable>
+struct ColumnToArrowConverter<PT, AT, is_nullable, ConvDecimalGuard<PT, AT>> {
+    using StarRocksCppType = RunTimeCppType<PT>;
+    using StarRocksColumnType = RunTimeColumnType<PT>;
+    using ArrowType = ArrowTypeIdToType<AT>;
+    using ArrowCppType = ArrowTypeIdToCppType<AT>;
+    using ArrowBuilderType = typename arrow::TypeTraits<ArrowType>::BuilderType;
+
+    static inline arrow::Decimal128 convert_datum(const StarRocksCppType& datum) {
+        int128_t value;
+        if constexpr (pt_is_decimalv2<PT>) {
+            value = datum.value();
+        } else if constexpr (pt_is_decimal<PT>) {
+            value = datum;
+        } else {
+            static_assert(pt_is_decimalv2<PT> || pt_is_decimal<PT>, "Illegal PrimitiveType");
+        }
+        int64_t high = value >> 64;
```

Comment:
AFAIK, someone in the community is working on porting StarRocks to ARM. Should we care about little endian and big endian from now?

Reply:
no need, shift operations are immune